[IMP][sale_order_line_packaging_qty] Sell product / unit or packaging

Tonow-c2c opened the pull request:
**Context:**

We want to be able to only sell by packaging some product (forbidden to
 sell by unit). Also, we want to choose which packaging is eligible to
 sell.

**Usage**

When I encode a sale order line, I must fill up the packaging and
related quantity if the product is set as "Only sell by packaging".
I cannot enter a quantity in the standard Odoo field quantity that
is not a multiple of a sellable packaging
(product.packaging can be sold = True).

Is the product is not configured as "Only sell by packaging",
then I am allowed to fill up a line with no packaging.

On every packaging I can chose if it is sellable or not. By default it is.

I can chose at product level if i restrict the quantity to sell
to existing sellable packaging.

**Dev**

* Add a check box in product "Only sell by packaging multiple" (default False), aside of "Can be sold" in the header
* Add a boolean on the packaging "Can be sold" (default True)
* On the list of available packaging in a SO line, only display packaging with "can be sold = True"
* When a SO line is encoded *with a product that can only be sold by packaging*:
    * The quantity field cannot contain a quantity that is not a multiple of a sellable packaging (can be sold = True) -> raise a warning (triggered on the on_change please, not create/write as we still want to be able to save it e.g. in case of an EDI order) -> "The quantity is not a multiple of a defined sellable packaging. Please complete the packaging first."
    * On create / write method, *if packaging and packaging quantity is empty*, fulfill the info if a multiple of an existing sellable packaging (can be sold = True) is found
        * e.g. if my product is sold by packaging only and I have packaging of box: 10, Karton: 200, and pallet:1000, if the quantity entered is 400, I expect the system to complete the packaging with Karton and packaging quantity with 2
        * e.g. if my product is sold by packaging only and I have packaging of box: 10, Karton: 200, and pallet:1000, if the quantity entered is 420, I expect the system to complete the packaging with box and packaging quantity with 42
        * e.g.. if my product is sold by packaging only and I have packaging of box: 10, Karton: 200, and pallet:1000, if the quantity entered is 405, I expect the system to leave it empty
    * Make the message "You must define a package before setting a quantity of said package." a user warning and not a blocking point AND only raise this warning for product that have "Only sell by packaging multiple = True"


-----


![OCA 2020-06-25 08-40](https://user-images.githubusercontent.com/32262135/85669657-5be60a00-b6c0-11ea-9ffe-5d8dc1046ede.gif)

rousseldenis commented on sale_order_line_packaging_qty/tests/test_sale_order_line_packaging_qty.py:
```diff
@@ -30,3 +30,58 @@ def test_product_packaging_qty(self):
+        order_line.write({"product_uom_qty": self.packaging.qty * 2})
+        self.assertFalse(order_line._onchange_product_uom_qty())
+
+    # def test_auto_fill_packaging(self):
```
Remove what you don't use

Tonow-c2c replied:
I will fix it's it's reason why I added on WIP on the description :wink:

rousseldenis commented:
Makes sense to me

Tonow-c2c committed:
[IMP][sale_order_line_packaging_qty] Sell product / unit or packaging

**Context:**

We want to be able to only sell by packaging some product (forbidden to
 sell by unit). Also, we want to choose which packaging is eligible to
 sell.

**Usage**

When I encode a sale order line, I must fill up the packaging and
related quantity if the product is set as "Only sell by packaging".
I cannot enter a quantity in the standard Odoo field quantity that
is not a multiple of a sellable packaging
(product.packaging can be sold = True).

Is the product is not configured as "Only sell by packaging",
then I am allowed to fill up a line with no packaging.

On every packaging I can chose if it is sellable or not. By default it is.

I can chose at product level if i restrict the quantity to sell
to existing sellable packaging.

**Dev**

* Add a check box in product "Only sell by packaging multiple" (default False), aside of "Can be sold" in the header
* Add a boolean on the packaging "Can be sold" (default True)
* On the list of available packaging in a SO line, only display packaging with "can be sold = True"
* When a SO line is encoded *with a product that can only be sold by packaging*:
    * The quantity field cannot contain a quantity that is not a multiple of a sellable packaging (can be sold = True) -> raise a warning (triggered on the on_change please, not create/write as we still want to be able to save it e.g. in case of an EDI order) -> "The quantity is not a multiple of a defined sellable packaging. Please complete the packaging first."
    * On create / write method, *if packaging and packaging quantity is empty*, fulfill the info if a multiple of an existing sellable packaging (can be sold = True) is found
        * e.g. if my product is sold by packaging only and I have packaging of box: 10, Karton: 200, and pallet:1000, if the quantity entered is 400, I expect the system to complete the packaging with Karton and packaging quantity with 2
        * e.g. if my product is sold by packaging only and I have packaging of box: 10, Karton: 200, and pallet:1000, if the quantity entered is 420, I expect the system to complete the packaging with box and packaging quantity with 42
        * e.g.. if my product is sold by packaging only and I have packaging of box: 10, Karton: 200, and pallet:1000, if the quantity entered is 405, I expect the system to leave it empty
    * Make the message "You must define a package before setting a quantity of said package." a user warning and not a blocking point AND only raise this warning for product that have "Only sell by packaging multiple = True"

i-vyshnevska commented on sale_order_line_packaging_qty/models/product_product.py:
```diff
@@ -0,0 +1,18 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
copyrights

i-vyshnevska commented on sale_order_line_packaging_qty/models/product_product.py:
```diff
@@ -0,0 +1,18 @@
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    def _which_pack_multiple(self, qty):
```
a comment?

simahawk replied:
a docstring! :)

i-vyshnevska commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -86,5 +87,69 @@ def _onchange_product_uom_qty(self):
+        return is_pack_multiple
+
+    def write(self, vals):
+        # Fill the packaging if they is empty and the quantity is a multiple
```
```suggestion
        # Fill the packaging if they are empty and the quantity is a multiple
```

i-vyshnevska commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -86,5 +87,69 @@ def _onchange_product_uom_qty(self):
+    @api.model
+    def create(self, vals):
+
+        # Fill the packaging if they is empty and the quantity is a multiple
```
```suggestion
        # Fill the packaging if they are empty and the quantity is a multiple
```

TDu commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -86,5 +87,69 @@ def _onchange_product_uom_qty(self):
+    def _check_qty_is_pack_multiple(self):
+        """ Check only for product with only_sell_by_pack
+        """
+        # self.product_uom_qty != 1 cause is the defaut value
```
Is this a todo that should be implemented ?

Tonow-c2c replied:
yes should be remove it's was comment for me in the dev phase sorry for the noise

TDu commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -86,5 +87,69 @@ def _onchange_product_uom_qty(self):
+        return {}
+
+    def _is_pack_multiple(self):
+        is_pack_multiple = False
```
This function is really closed to `def _which_pack_multiple(self, qty)` on `product.product` 
I wonder if you could just do something like this
```suggestion
        return bool(self.product_id._which_pack_multiple(self.product_uom_qty))
```

simahawk commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -86,5 +87,69 @@ def _onchange_product_uom_qty(self):
+                    "message": _(
+                        "For the product {prod}\n"
+                        "The {qty} is not the multiple of any pack.\n"
+                        "Please add a package".format(
```
translation broken if you format it on the string. Use `format` on the gettext object: `_(...).format(...)`

simahawk commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -86,5 +87,69 @@ def _onchange_product_uom_qty(self):
+        # and we dont want to have this warning when we had the product
+        if self.product_id.only_sell_by_pack:
+            if not self._is_pack_multiple():
+                warning_mess = {
```
"mess" != "msg" :smile:

simahawk commented on sale_order_line_packaging_qty/models/product_template.py:
```diff
@@ -0,0 +1,9 @@
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    only_sell_by_pack = fields.Boolean(string="Only sell by packaging", default=False,)
```
```suggestion
    sell_only_by_packaging = fields.Boolean(string="Only sell by packaging", default=False,)
```

jgrandguillaume reviewed (approved):
Unless you fix what @simahawk asked I'm good, thx !

Tonow-c2c commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -86,5 +87,69 @@ def _onchange_product_uom_qty(self):
+    def _check_qty_is_pack_multiple(self):
+        """ Check only for product with only_sell_by_pack
+        """
+        # self.product_uom_qty != 1 cause is the defaut value
```
```suggestion
```

simahawk commented:
@Tonow-c2c thanks! Please squash! :)

jgrandguillaume commented:
/ocabot merge minor

grindtildeath commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -86,5 +87,60 @@ def _onchange_product_uom_qty(self):
+    def _is_pack_multiple(self):
+        return bool(self.product_id._which_pack_multiple(self.product_uom_qty))
+
+    def write(self, vals):
+        # Fill the packaging if they are empty and the quantity is a multiple
+        for line in self:
+            product_uom_qty = vals.get("product_uom_qty")
+            product_packaging = vals.get("product_packaging")
+            if line.product_id.sell_only_by_packaging and (
+                not line.product_packaging
+                or ("product_packaging" in vals and not product_packaging)
+            ):
+                pack_multiple = line.product_id._which_pack_multiple(product_uom_qty)
+                if pack_multiple:
+                    vals.update({"product_packaging": pack_multiple.id})
```
How could this go unnoticed? :astonished: 

1. If `product_id` is defined in `vals` accessing `line.product_id` to select a packaging would use packaging of a product that is not the one written on the lines.
2. If `product_id` is not defined in `vals`, the for loop on the lines will redefine `product_packaging` on each iteration that finds a packaging, and the value written on all the lines will be the last packaging to be found.

grindtildeath reviewed (commented):
I have two issues with this PR:
- Mix of multiple features into the same module
- Buggy code (cf below)

That's why I did open https://github.com/OCA/sale-workflow/pull/1191 to revert this.
